lib/cretonne/src/simple_gvn.rs

```diff
@@ -0,0 +1,133 @@
+
+use flowgraph::{ControlFlowGraph};
+use dominator_tree::DominatorTree;
+use ir::{Cursor, DataFlowGraph, InstructionData, Function, Inst, Opcode};
+use std::hash::{Hash, Hasher};
+use std::collections::HashMap;
+
+/// Test whether the given opcode is unsafe to even consider for GVN.
+fn trivially_unsafe_for_gvn(opcode: Opcode) -> bool {
+    opcode.is_call() || opcode.is_branch() || opcode.is_terminator() ||
+    opcode.is_return() || opcode.can_trap()
+}
+
+#[derive(Debug)]
+struct VisibleInst(InstructionData);
+
+impl Hash for VisibleInst {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // TODO: Is a better way to do this? Should we just make InstructionData
+        // implement Hash itself?
+        use ir::instructions::InstructionData::*;
+        match self.0 {
+            Nullary { .. } => {},
+            Unary { arg, .. } => arg.hash(state),
+            Binary { args, .. } => args.hash(state),
+            Ternary { args, .. } => args.hash(state),
+            _ => {
+                panic!("TODO: implement hashing for more variants.");
+            },
+        }
+    }
+}
+
+impl PartialEq for VisibleInst {
+    fn eq(&self, other: &VisibleInst) -> bool {
+        if self.0.opcode() != other.0.opcode() {
+            return false;
+        }
+
+        // TODO: Surely there is a better way to do this? Should we provide a
+        // PartialEq implementation for InstructionData? Note that it isn't as
+        // simple as just using the default implementation, since
+        // InstructionData contains things like floating-point literals which
+        // don't have Eq defined.
+        use ir::instructions::InstructionData::*;
+        match self.0 {
+            Nullary { .. } => true,
+            Unary { arg, .. } => match other.0 {
+                Unary { arg: other_arg, .. } => arg == other_arg,
+                _ => panic!("we checked the opcode so this shouldn't happen"),
+            },
+            Binary { args, .. } => match other.0 {
+                Binary { args: other_args, .. } => args == other_args,
+                _ => panic!("we checked the opcode so this shouldn't happen"),
+            },
+            Ternary { args, .. } => match other.0 {
+                Ternary { args: other_args, .. } => args == other_args,
+                _ => panic!("we checked the opcode so this shouldn't happen"),
+            },
+            _ => {
+                // TODO: Implement eq for more variants.
+                false
+            },
+        }
+    }
+}
+impl Eq for VisibleInst {}
+
+/// Replace the values of new_inst with the values of old_inst.
+///
+fn replace_values(dfg: &mut DataFlowGraph, new_inst: Inst, old_inst: Inst) {
+    let mut result = 0;
+    let new_results = dfg.detach_results(new_inst);
+    while let Some(new_val) = new_results.get(result, &dfg.value_lists) {
+        let old_val = dfg.inst_results(old_inst)[result];
+        dfg.change_to_alias(new_val, old_val);
+        result += 1;
+    }
```

Comment:
Sometimes the borrow checker is really annoying.

If you implement this as a method on `DataFlowGraph`, the borrow checker will be able to understand that `dfg.value_lists` and `dfg.values` don't alias, and you can iterate over the two slices of results instead.

Rust does not fully support functional decomposition...

```rust
for (&new, &old) in new_results.as_slice(...).iter().zip(self.results[old_inst].as_slice(...)) {
   // inline `change_to_alias()` here...
}
```